pkg/pillar/cmd/zedrouter/zedrouter.go

```diff
@@ -1242,59 +1263,72 @@ func findBridge(bridgeName string) (*netlink.Bridge, error) {
 }
 
 // XXX Need additional logic for IPv6 underlays.
-func getUlAddrs(ctx *zedrouterContext,
-	ifnum int, appNum int,
-	status *types.UnderlayNetworkStatus,
-	netInstStatus *types.NetworkInstanceStatus) (string, string, error) {
-
-	log.Functionf("getUlAddrs(%d/%d)\n", ifnum, appNum)
-
-	bridgeIPAddr := ""
-	appIPAddr := ""
-
-	// Allocate bridgeIPAddr based on BridgeMac
-	log.Functionf("getUlAddrs(%d/%d for %s) bridgeMac %s\n",
-		ifnum, appNum, netInstStatus.UUID.String(),
-		status.BridgeMac.String())
+func getUlAddrs(ctx *zedrouterContext, ifnum int,
+	netInstStatus *types.NetworkInstanceStatus,
+	ulStatus *types.UnderlayNetworkStatus, appID uuid.UUID) (string, error) {
 	var err error
-	var addr string
-	addr, err = lookupOrAllocateIPv4(ctx, netInstStatus,
-		status.BridgeMac)
+	var mac net.HardwareAddr
+	ipAddr := ""
+	appNum := types.AppNumInvalid
+	networkID := netInstStatus.UUID
+	log.Functionf("getUlAddrs(%d) for %s/%s\n", ifnum,
+		networkID.String(), appID.String())
+	mac, err = net.ParseMAC(ulStatus.Mac)
 	if err != nil {
-		log.Errorf("getUlAddrs: Bridge IP address allocation failed %s\n", err)
-		return bridgeIPAddr, appIPAddr, err
+		errStr := fmt.Sprintf("getUlAddrs(%s/%s): parse Mac fail: %v\n",
+			networkID.String(), appID.String(), err)
+		log.Errorf(errStr)
+		return "", errors.New(errStr)
+	}
+	// for static IP Address
+	if ulStatus.AppIPAddr != nil {
+		ipAddr = ulStatus.AppIPAddr.String()
+		if netInstStatus.DhcpRange.Contains(ulStatus.AppIPAddr) {
+			errStr := fmt.Sprintf("getUlAddrs(%s/%s): Dhcp Range contains the static IP Address %s\n",
+				networkID.String(), appID.String(), ipAddr)
+			log.Errorf(errStr)
+			return "", errors.New(errStr)
+		}
 	} else {
-		bridgeIPAddr = addr
-	}
-
-	if status.AppIPAddr != nil {
-		// Static IP assignment case.
-		// Note that appIPAddr can be in a different subnet.
-		// Assumption is that the config specifies a gateway/router
-		// in the same subnet as the static address.
-		appIPAddr = status.AppIPAddr.String()
-		recordIPAssignment(ctx, netInstStatus, status.AppIPAddr,
-			status.Mac)
-	} else if status.Mac != "" {
-		// XXX or change type of VifInfo.Mac to avoid parsing?
-		var mac net.HardwareAddr
-		mac, err = net.ParseMAC(status.Mac)
+		appNum, err = appNumOnUNetAlloc(ctx, networkID, appID,
```

Comment:
It is utterly confusing to call this variable (and also where it is used as an argument) "appNum" when we already have an appNum. This needs to be called appNumOnNet or appNumOnUnet (the "underlay" aka "U" isn't needed"). But appNum will cause confusion.